### PR TITLE
Added anchor support

### PR DESCRIPTION
**Pull request by judax**
Initial implementation for Anchors in ARCore. All the basic structures have been added but there is still testing to be done. Mainly because there seems to be a crash in the event handling.

### CODE REVIEW

File: android_webview/test/shell/tango/jni/AnchorManager.cpp [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+}
+
+uint32_t Anchor::getIdentifier() const
+{
+  return identifier;
+}
+
+const float* Anchor::getModelMatrix() const
+{
+  return (const float*)(&anchorModelMatrix);
```
Comment (lincolnfrog) [inline android_webview/test/shell/tango/jni/AnchorManager.cpp]: return glm::value_ptr(anchorModelMatrix);

File: android_webview/test/shell/tango/jni/AnchorManager.cpp [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+  std::shared_ptr<Anchor> anchor(new Anchor(timestamp,
+                                            cameraModelMatrix,
+                                            anchorModelMatrix));
+  anchors[anchor->getIdentifier()] = anchor;
+  return anchor;
+}
+
+void AnchorManager::removeAnchor(uint32_t identifier)
+{
+  std::unordered_map<uint32_t, std::shared_ptr<Anchor>>::iterator it =
```
Comment (lincolnfrog) [inline android_webview/test/shell/tango/jni/AnchorManager.cpp]: auto it = ...

File: android_webview/test/shell/tango/jni/AnchorManager.cpp [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+  return anchor;
+}
+
+void AnchorManager::removeAnchor(uint32_t identifier)
+{
+  std::unordered_map<uint32_t, std::shared_ptr<Anchor>>::iterator it = 
+    anchors.find(identifier);
+  if (it != anchors.end())
+  {
+    anchors.erase(it);
```
Comment (lincolnfrog) [inline android_webview/test/shell/tango/jni/AnchorManager.cpp]: erase actually supports erase-by-key: http://www.cplusplus.com/reference/unordered_map/unordered_map/erase/

This whole method can be (I believe) "anchors.erase(identifier);"

File: android_webview/test/shell/tango/jni/AnchorManager.cpp [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+{
+  anchors.clear();
+}
+
+std::vector<std::shared_ptr<Anchor>> AnchorManager::update(
+  double newTimestamp, const float* newCameraModelMatrix)
+{
+  std::vector<std::shared_ptr<Anchor>> updatedAnchors;
+  updatedAnchors.reserve(anchors.size());
+  std::unordered_map<uint32_t, std::shared_ptr<Anchor>>::const_iterator it =
```
Comment (lincolnfrog) [inline android_webview/test/shell/tango/jni/AnchorManager.cpp]: auto it

File: android_webview/test/shell/tango/jni/TangoHandler.h [hunk 5]
```diff
@@ -77,7 +82,8 @@ class TangoHandler {
 
   ~TangoHandler();
 
-  void onCreate(JNIEnv* env, jobject activity, int activityOrientation, int sensorOrientation);
+  void onCreate(JNIEnv* env, jobject activity, 
+  							int activityOrientation, int sensorOrientation);
```
Comment (lincolnfrog) [inline android_webview/test/shell/tango/jni/TangoHandler.h]: this is all tabs :(

File: android_webview/test/shell/tango/jni/TangoHandler.h [hunk 6]
```diff
@@ -89,9 +95,14 @@ class TangoHandler {
   bool getProjectionMatrix(float near, float far, float* projectionMatrix);
   bool hitTest(float x, float y, std::vector<Hit>& hits);
   bool getPlaneDeltas(PlaneDeltas& planeDeltas);
+  std::shared_ptr<Anchor> createAnchor(
+  	const float* anchorModelMatrix);
```
Comment (lincolnfrog) [inline android_webview/test/shell/tango/jni/TangoHandler.h]: tab

File: android_webview/test/shell/tango/jni/TangoHandler.h [hunk 7]
```diff
@@ -64,6 +61,14 @@ class PlaneDeltas
 	std::vector<long> removed;
 };
 
+class TangoHandlerEventListener
+{
+public:
+  virtual void anchorsUpdated(
+  	const std::vector<std::shared_ptr<Anchor>>& anchors) = 0;
```
Comment (lincolnfrog) [inline android_webview/test/shell/tango/jni/TangoHandler.h]: tab

File: device/vr/android/tango/tango_vr_device.cc [hunk 8]
```diff
@@ -269,6 +281,30 @@ mojom::VRPlaneDeltasPtr TangoVRDevice::GetPlaneDeltas() {
   return mojomPlaneDeltas;
 }
 
+mojom::VRAnchorPtr TangoVRDevice::CreateAnchor(
+    const std::vector<float>& modelMatrix) 
+{
+  if (!TangoHandler::getInstance()->isConnected()) 
+  {
```
Comment (lincolnfrog) [inline device/vr/android/tango/tango_vr_device.cc:289]: shouldn't this be on the previous line? I believe that is Google c++ style (I know its js and java style).

File: device/vr/android/tango/tango_vr_device.cc [hunk 9]
```diff
@@ -269,6 +281,30 @@ mojom::VRPlaneDeltasPtr TangoVRDevice::GetPlaneDeltas() {
...
+    return nullptr;
+  }
+  mojom::VRAnchorPtr mojomAnchor = nullptr;
+  std::shared_ptr<Anchor> anchor = TangoHandler::getInstance()->
+    createAnchor((const float*)(&(modelMatrix[0])));
+  if (!anchor)
+  {
```
Comment (lincolnfrog) [inline device/vr/android/tango/tango_vr_device.cc:296]: { should be on the previous line according to Google c++ style (go/cstyle)

File: device/vr/android/tango/tango_vr_device.cc [hunk 10]
```diff
@@ -309,4 +345,20 @@ void TangoVRDevice::UpdateLayerBounds(mojom::VRLayerBoundsPtr left_bounds,
   // delegate_->UpdateWebVRTextureBounds(left_gvr_bounds, right_gvr_bounds);
 }
 
+void TangoVRDevice::anchorsUpdated(
+  const std::vector<std::shared_ptr<Anchor>>& anchors)
+{
+  std::vector<mojom::VRAnchorPtr> mojomAnchors;
+  mojomAnchors.resize(anchors.size());
+  for (size_t i = 0; i < anchors.size(); i++)
+  {
```
Comment (lincolnfrog) [inline device/vr/android/tango/tango_vr_device.cc:355]: { should be on the previous line according to Google c++ style (go/cstyle)

File: device/vr/android/tango/tango_vr_device.cc [hunk 11]
```diff
@@ -309,4 +345,20 @@ void TangoVRDevice::UpdateLayerBounds(mojom::VRLayerBoundsPtr left_bounds,
   // delegate_->UpdateWebVRTextureBounds(left_gvr_bounds, right_gvr_bounds);
 }
 
+void TangoVRDevice::anchorsUpdated(
+  const std::vector<std::shared_ptr<Anchor>>& anchors)
+{
```
Comment (lincolnfrog) [inline device/vr/android/tango/tango_vr_device.cc:351]: { should be on the previous line according to Google c++ style (go/cstyle)

File: device/vr/android/tango/tango_vr_device.cc [hunk 12]
```diff
@@ -269,6 +281,30 @@ mojom::VRPlaneDeltasPtr TangoVRDevice::GetPlaneDeltas() {
   return mojomPlaneDeltas;
 }
 
+mojom::VRAnchorPtr TangoVRDevice::CreateAnchor(
+    const std::vector<float>& modelMatrix) 
+{
```
Comment (lincolnfrog) [inline device/vr/android/tango/tango_vr_device.cc:287]: { should be on the previous line according to Google c++ style (go/cstyle)

File: device/vr/android/tango/tango_vr_device.cc [hunk 13]
```diff
@@ -239,6 +239,18 @@ static mojom::VRPlanePtr CreateMojomPlane(Plane& plane) {
   return result;
 }
 
+static mojom::VRAnchorPtr CreateMojomAnchor(const std::shared_ptr<Anchor>& anchor)
+{
```
Comment (lincolnfrog) [inline device/vr/android/tango/tango_vr_device.cc:244]: { should be on the previous line according to Google c++ style (go/cstyle)

File: android_webview/test/shell/tango/jni/AnchorManager.cpp [hunk 14]
```diff
@@ -0,0 +1,89 @@
...
+    anchors.begin();
+  for (; it != anchors.end(); it++) {
+    std::shared_ptr<Anchor> anchor = it->second;
+    if (anchor->update(newTimestamp, newCameraModelMatrix))
+    {
```
Comment (lincolnfrog) [inline android_webview/test/shell/tango/jni/AnchorManager.cpp]: { should be on the previous line according to Google c++ style (go/cstyle)

File: android_webview/test/shell/tango/jni/AnchorManager.cpp [hunk 15]
```diff
@@ -0,0 +1,89 @@
...
+}
+
+void AnchorManager::removeAllAnchors()
+{
+  anchors.clear();
+}
+
+std::vector<std::shared_ptr<Anchor>> AnchorManager::update(
+  double newTimestamp, const float* newCameraModelMatrix)
+{
```
Comment (lincolnfrog) [inline android_webview/test/shell/tango/jni/AnchorManager.cpp]: { should be on the previous line according to Google c++ style (go/cstyle)

File: android_webview/test/shell/tango/jni/AnchorManager.cpp [hunk 16]
```diff
@@ -0,0 +1,89 @@
...
+  std::unordered_map<uint32_t, std::shared_ptr<Anchor>>::iterator it = 
+    anchors.find(identifier);
+  if (it != anchors.end())
+  {
+    anchors.erase(it);
+  }
+}
+
+void AnchorManager::removeAllAnchors()
+{
```
Comment (lincolnfrog) [inline android_webview/test/shell/tango/jni/AnchorManager.cpp]: { should be on the previous line according to Google c++ style (go/cstyle)

File: android_webview/test/shell/tango/jni/AnchorManager.cpp [hunk 17]
```diff
@@ -0,0 +1,89 @@
...
+  anchors[anchor->getIdentifier()] = anchor;
+  return anchor;
+}
+
+void AnchorManager::removeAnchor(uint32_t identifier)
+{
+  std::unordered_map<uint32_t, std::shared_ptr<Anchor>>::iterator it = 
+    anchors.find(identifier);
+  if (it != anchors.end())
+  {
```
Comment (lincolnfrog) [inline android_webview/test/shell/tango/jni/AnchorManager.cpp]: { should be on the previous line according to Google c++ style (go/cstyle)

File: android_webview/test/shell/tango/jni/AnchorManager.cpp [hunk 18]
```diff
@@ -0,0 +1,89 @@
...
+{
+  std::shared_ptr<Anchor> anchor(new Anchor(timestamp,
+                                            cameraModelMatrix,
+                                            anchorModelMatrix));
+  anchors[anchor->getIdentifier()] = anchor;
+  return anchor;
+}
+
+void AnchorManager::removeAnchor(uint32_t identifier)
+{
```
Comment (lincolnfrog) [inline android_webview/test/shell/tango/jni/AnchorManager.cpp]: { should be on the previous line according to Google c++ style (go/cstyle)

File: android_webview/test/shell/tango/jni/AnchorManager.cpp [hunk 19]
```diff
@@ -0,0 +1,89 @@
...
+const float* Anchor::getModelMatrix() const
+{
+  return (const float*)(&anchorModelMatrix);
+}
+
+// ===============================================
+
+std::shared_ptr<Anchor> AnchorManager::createAnchor(double timestamp,
+  const float* cameraModelMatrix, const float* anchorModelMatrix)
+{
```
Comment (lincolnfrog) [inline android_webview/test/shell/tango/jni/AnchorManager.cpp]: { should be on the previous line according to Google c++ style (go/cstyle)

File: android_webview/test/shell/tango/jni/AnchorManager.cpp [hunk 20]
```diff
@@ -0,0 +1,89 @@
...
+  return needsUpdate;
+}
+
+uint32_t Anchor::getIdentifier() const
+{
+  return identifier;
+}
+
+const float* Anchor::getModelMatrix() const
+{
```
Comment (lincolnfrog) [inline android_webview/test/shell/tango/jni/AnchorManager.cpp]: { should be on the previous line according to Google c++ style (go/cstyle)

File: android_webview/test/shell/tango/jni/AnchorManager.cpp [hunk 21]
```diff
@@ -0,0 +1,89 @@
...
+    // TODO: Why store the cameraModelMatrix? A local variable would do it.
+    memcpy(glm::value_ptr(cameraModelMatrix), newCameraModelMatrix, 
+           sizeof(glm::mat4));
+    anchorModelMatrix = cameraModelMatrix * anchorModelMatrixRelativeToCamera;
+  }
+  return needsUpdate;
+}
+
+uint32_t Anchor::getIdentifier() const
+{
```
Comment (lincolnfrog) [inline android_webview/test/shell/tango/jni/AnchorManager.cpp]: { should be on the previous line according to Google c++ style (go/cstyle)

File: android_webview/test/shell/tango/jni/AnchorManager.cpp [hunk 22]
```diff
@@ -0,0 +1,89 @@
...
+  glm::mat4 inverseCameraModelMatrix = glm::inverse(this->cameraModelMatrix);
+  anchorModelMatrixRelativeToCamera =
+    inverseCameraModelMatrix * this->anchorModelMatrix;
+}
+
+bool Anchor::update(double newTimestamp, const float* newCameraModelMatrix)
+{
+  bool needsUpdate = newTimestamp <= timestamp;
+  if (needsUpdate)
+  {
```
Comment (lincolnfrog) [inline android_webview/test/shell/tango/jni/AnchorManager.cpp]: { should be on the previous line according to Google c++ style (go/cstyle)

File: android_webview/test/shell/tango/jni/AnchorManager.cpp [hunk 23]
```diff
@@ -0,0 +1,89 @@
...
+         sizeof(glm::mat4));
+  memcpy(glm::value_ptr(this->anchorModelMatrix), anchorModelMatrix,
+         sizeof(glm::mat4));
+  glm::mat4 inverseCameraModelMatrix = glm::inverse(this->cameraModelMatrix);
+  anchorModelMatrixRelativeToCamera =
+    inverseCameraModelMatrix * this->anchorModelMatrix;
+}
+
+bool Anchor::update(double newTimestamp, const float* newCameraModelMatrix)
+{
```
Comment (lincolnfrog) [inline android_webview/test/shell/tango/jni/AnchorManager.cpp]: { should be on the previous line according to Google c++ style (go/cstyle)

File: android_webview/test/shell/tango/jni/AnchorManager.cpp [hunk 24]
```diff
@@ -0,0 +1,89 @@
...
+#include "gtc/type_ptr.hpp"
+
+using namespace tango_chromium;
+
+uint32_t Anchor::identifierCounter = 0;
+
+Anchor::Anchor(double timestamp, const float* cameraModelMatrix,
+               const float* anchorModelMatrix):
+                 identifier(identifierCounter++), timestamp(timestamp)
+{
```
Comment (lincolnfrog) [inline android_webview/test/shell/tango/jni/AnchorManager.cpp]: { should be on the previous line according to Google c++ style (go/cstyle)

File: device/vr/vr_device.cc [hunk 25]
```diff
@@ -51,6 +51,15 @@ void VRDevice::OnChanged() {
     display->client()->OnChanged(vr_device_info.Clone());
 }
 
+void VRDevice::OnAnchorsUpdated(std::vector<mojom::VRAnchorPtr> anchors)
+{
```
Comment (lincolnfrog) [inline device/vr/vr_device.cc]: { should be on the previous line according to Google c++ style (go/cstyle)

File: device/vr/vr_device.cc [hunk 26]
```diff
@@ -51,6 +51,15 @@ void VRDevice::OnChanged() {
...
+  
+  VLOG(0) << "JUDAX: VRDevice::OnAnchorsUpdated -> displays_.size() = " << displays_.size() << ", anchors.size() = " << anchors.size();
+
+  for (const auto& display : displays_)
```
Comment (lincolnfrog) [inline device/vr/vr_device.cc]: always surround blocks with {} according to Google style

File: third_party/WebKit/Source/modules/vr/VRAnchor.cpp [hunk 27]
```diff
@@ -0,0 +1,32 @@
+// Copyright 2016 The Chromium Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#include "modules/vr/VRAnchor.h"
+
+namespace blink {
+
+VRAnchor::VRAnchor()
+{
```
Comment (lincolnfrog) [inline third_party/WebKit/Source/modules/vr/VRAnchor.cpp]: { should be on the previous line according to Google c++ style (go/cstyle)

File: third_party/WebKit/Source/modules/vr/VRAnchor.cpp [hunk 28]
```diff
@@ -0,0 +1,32 @@
...
+    m_modelMatrix = DOMFloat32Array::create(16);
+}
+
+void VRAnchor::setAnchor(const device::mojom::blink::VRAnchorPtr& anchorPtr)
+{
```
Comment (lincolnfrog) [inline third_party/WebKit/Source/modules/vr/VRAnchor.cpp]: { should be on the previous line according to Google c++ style (go/cstyle)

File: third_party/WebKit/Source/modules/vr/VRAnchor.cpp [hunk 29]
```diff
@@ -0,0 +1,32 @@
...
+
+    m_identifier = anchorPtr->identifier;
+
+    for (size_t i = 0; i < 16; i++) {
+      m_modelMatrix->data()[i] = (float)anchorPtr->modelMatrix[i];
+    }
+}
+
+DEFINE_TRACE(VRAnchor)
+{
```
Comment (lincolnfrog) [inline third_party/WebKit/Source/modules/vr/VRAnchor.cpp]: { should be on the previous line according to Google c++ style (go/cstyle)

File: third_party/WebKit/Source/modules/vr/VRAnchorEvent.cpp [hunk 30]
```diff
@@ -0,0 +1,44 @@
...
+#include "modules/vr/VRAnchorEvent.h"
+
+#include "modules/vr/VRAnchor.h"
+
+namespace blink {
+
+VRAnchorEvent::VRAnchorEvent() {}
+
+VRAnchorEvent::VRAnchorEvent(const AtomicString& type,
+                           bool canBubble,
```
Comment (lincolnfrog) [inline third_party/WebKit/Source/modules/vr/VRAnchorEvent.cpp]: indentation one more space for this whole block?

File: third_party/WebKit/Source/modules/vr/VRAnchorEvent.cpp [hunk 31]
```diff
@@ -0,0 +1,44 @@
...
+                           bool cancelable,
+                           VRDisplay* display,
+                           HeapVector<Member<VRAnchor>> anchors)
+    : Event(type, canBubble, cancelable),
+      m_display(display),
+      m_anchors(anchors) {}
+
+VRAnchorEvent::VRAnchorEvent(const AtomicString& type,
+                           const VRAnchorEventInit& initializer)
```
Comment (lincolnfrog) [inline third_party/WebKit/Source/modules/vr/VRAnchorEvent.cpp]: one more space indent

File: third_party/WebKit/Source/modules/vr/VRAnchorEvent.cpp [hunk 32]
```diff
@@ -0,0 +1,44 @@
...
+    : Event(type, initializer) {
+  if (initializer.hasDisplay())
+    m_display = initializer.display();
```
Comment (lincolnfrog) [inline third_party/WebKit/Source/modules/vr/VRAnchorEvent.cpp]: always use {} for blocks

File: third_party/WebKit/Source/modules/vr/VRAnchorEvent.cpp [hunk 33]
```diff
@@ -0,0 +1,44 @@
...
+
+  if (initializer.hasAnchors())
```
Comment (lincolnfrog) [inline third_party/WebKit/Source/modules/vr/VRAnchorEvent.cpp:28]: always use {} for blocks

File: third_party/WebKit/Source/modules/vr/VRAnchorEvent.h [hunk 34]
```diff
@@ -0,0 +1,57 @@
...
+  }
+
+  static VRAnchorEvent* create(const AtomicString& type,
+                              const VRAnchorEventInit& initializer) {
+    return new VRAnchorEvent(type, initializer);
+  }
+
+  ~VRAnchorEvent() override;
+
+  VRDisplay* display() const { return m_display.get(); }
```
Comment (lincolnfrog) [inline third_party/WebKit/Source/modules/vr/VRAnchorEvent.h:35]: I am pretty sure you aren't supposed to have spaces within {} blocks like this according to Google style (at least for python, can't find it for c, but check go/cstyle and see if you can find it).

File: third_party/WebKit/Source/modules/vr/VRAnchorEvent.h [hunk 35]
```diff
@@ -0,0 +1,57 @@
...
+  HeapVector<Member<VRAnchor>> anchors() const { return m_anchors; }
+
+  const AtomicString& interfaceName() const override;
+
+  DECLARE_VIRTUAL_TRACE();
+
+ private:
+  VRAnchorEvent();
+  VRAnchorEvent(const AtomicString& type,
+               bool canBubble,
```
Comment (lincolnfrog) [inline third_party/WebKit/Source/modules/vr/VRAnchorEvent.h:45]: one more space indentation for this block

File: third_party/WebKit/Source/modules/vr/VRDisplay.cpp [hunk 36]
```diff
@@ -316,6 +318,63 @@ VRPassThroughCamera* VRDisplay::getPassThroughCamera()
   return m_passThroughCamera;
 }
 
+VRAnchor* VRDisplay::createAnchor(WTF::Vector<float>& modelMatrix)
+{
+  if (!m_display)
+    return nullptr;
```
Comment (lincolnfrog) [inline third_party/WebKit/Source/modules/vr/VRDisplay.cpp]: always use {} around blocks

File: third_party/WebKit/Source/modules/vr/VRDisplay.cpp [hunk 37]
```diff
@@ -316,6 +318,63 @@ VRPassThroughCamera* VRDisplay::getPassThroughCamera()
   return m_passThroughCamera;
 }
 
+VRAnchor* VRDisplay::createAnchor(WTF::Vector<float>& modelMatrix)
+{
```
Comment (lincolnfrog) [inline third_party/WebKit/Source/modules/vr/VRDisplay.cpp]: { should be on the previous line according to Google c++ style (go/cstyle)

File: third_party/WebKit/Source/modules/vr/VRDisplay.cpp [hunk 38]
```diff
@@ -316,6 +318,63 @@ VRPassThroughCamera* VRDisplay::getPassThroughCamera()
...
+  m_anchors.push_back(anchor);
+
+  VLOG(0) << "JUDAX: VRDisplay::createAnchor -> Anchor created with identifier = " <<
+             anchor->identifier();
+
+  return anchor;
+}
+
+VRAnchor* VRDisplay::createAnchor(DOMFloat32Array* modelMatrix)
+{
```
Comment (lincolnfrog) [inline third_party/WebKit/Source/modules/vr/VRDisplay.cpp]: { should be on the previous line according to Google c++ style (go/cstyle)

File: third_party/WebKit/Source/modules/vr/VRDisplay.cpp [hunk 39]
```diff
@@ -316,6 +318,63 @@ VRPassThroughCamera* VRDisplay::getPassThroughCamera()
...
+  if (!m_display)
+    return nullptr;
```
Comment (lincolnfrog) [inline third_party/WebKit/Source/modules/vr/VRDisplay.cpp]: always use {} around blocks

File: third_party/WebKit/Source/modules/vr/VRDisplay.cpp [hunk 40]
```diff
@@ -316,6 +318,63 @@ VRPassThroughCamera* VRDisplay::getPassThroughCamera()
...
+  for (size_t i = 0; i < modelMatrixSize; i++) {
+    modelMatrixVector[i] = modelMatrix->data()[i];
+  }
+
+  // Use the other overloaded method to create an anchor using the Vector.
+  return createAnchor(modelMatrixVector);
+}
+
+void VRDisplay::removeAnchor(VRAnchor* anchor)
+{
```
Comment (lincolnfrog) [inline third_party/WebKit/Source/modules/vr/VRDisplay.cpp]: { should be on the previous line according to Google c++ style (go/cstyle)

File: third_party/WebKit/Source/modules/vr/VRDisplay.cpp [hunk 41]
```diff
@@ -316,6 +318,63 @@ VRPassThroughCamera* VRDisplay::getPassThroughCamera()
...
+  if (!m_display)
+    return;
```
Comment (lincolnfrog) [inline third_party/WebKit/Source/modules/vr/VRDisplay.cpp]: always use {} around blocks

File: third_party/WebKit/Source/modules/vr/VRDisplay.cpp [hunk 42]
```diff
@@ -316,6 +318,63 @@ VRPassThroughCamera* VRDisplay::getPassThroughCamera()
...
+
+  bool removed = false;
+  for (size_t i = 0; !removed && i < m_anchors.size(); i++) {
+    // NOTE: Should we compare the pointers directly?
+    if (m_anchors[i]->identifier() == anchor->identifier()) {
+      m_anchors.remove(i);
+      removed = true;
```
Comment (lincolnfrog) [inline third_party/WebKit/Source/modules/vr/VRDisplay.cpp]: any reason to not just break here? you get rid of a bool and an extra check. Are you expecting multiple entries with matching ids?

File: third_party/WebKit/Source/modules/vr/VRDisplay.cpp [hunk 43]
```diff
@@ -316,6 +318,63 @@ VRPassThroughCamera* VRDisplay::getPassThroughCamera()
...
+    }
+  }
+}
+
+HeapVector<Member<VRAnchor>> VRDisplay::getAnchors()
+{
```
Comment (lincolnfrog) [inline third_party/WebKit/Source/modules/vr/VRDisplay.cpp]: { should be on the previous line according to Google c++ style (go/cstyle)

File: third_party/WebKit/Source/modules/vr/VRDisplay.cpp [hunk 44]
```diff
@@ -792,6 +851,40 @@ void VRDisplay::OnChanged(device::mojom::blink::VRDisplayInfoPtr display) {
...
 
+void VRDisplay::OnAnchorsUpdated(
+    WTF::Vector<device::mojom::blink::VRAnchorPtr> mojomAnchors) {
+
+  // TODO: WTF::Vector does not have empty or IsEmpty?
+  // if (mojomAnchors.size() == 0) {
+  //   return;
+  // }
+
+  VLOG(0) << "JUDAX: VRDisplay::OnAnchorsUpdated -> mojomAnchors.size() = " << mojomAnchors.size();
```
Comment (lincolnfrog) [inline third_party/WebKit/Source/modules/vr/VRDisplay.cpp:858]: do you want to check this in?

File: third_party/tango/libtango_chromium/AnchorManager.h [hunk 45]
```diff
@@ -0,0 +1,66 @@
...
+#define LOGI(...) __android_log_print(ANDROID_LOG_INFO, LOG_TAG, __VA_ARGS__)
+#define LOGE(...) __android_log_print(ANDROID_LOG_ERROR, LOG_TAG, __VA_ARGS__)
+
+namespace tango_chromium
+{
+
+class AnchorManager;
+
+class Anchor
+{
```
Comment (lincolnfrog) [inline third_party/tango/libtango_chromium/AnchorManager.h]: { should be on the previous line according to Google c++ style (go/cstyle)

File: third_party/tango/libtango_chromium/AnchorManager.h [hunk 46]
```diff
@@ -0,0 +1,66 @@
...
+  double timestamp;
+  glm::mat4 cameraModelMatrix;
+  glm::mat4 anchorModelMatrix;
+  glm::mat4 anchorModelMatrixRelativeToCamera;
+
+  friend class AnchorManager;
+};
+
+class AnchorManager
+{
```
Comment (lincolnfrog) [inline third_party/tango/libtango_chromium/AnchorManager.h]: { should be on the previous line according to Google c++ style (go/cstyle)

File: third_party/tango/libtango_chromium/AnchorManager.h [hunk 47]
```diff
@@ -0,0 +1,66 @@
...
+public:
+  std::shared_ptr<Anchor> createAnchor(double timestamp, 
+                                    const float* cameraModelMatrix,
```
Comment (lincolnfrog) [inline third_party/tango/libtango_chromium/AnchorManager.h]: add a couple spaces to this block indent

File: third_party/tango/libtango_chromium/AnchorManager.h [hunk 48]
```diff
@@ -0,0 +1,66 @@
...
+  void removeAllAnchors();
+
+  /**
+  * Return: The list of updated anchors.
+  */
+  std::vector<std::shared_ptr<Anchor>> update(
+    double newTimestamp, const float* newCameraModelMatrix);
+  
+private:
+  std::unordered_map<uint32_t, std::shared_ptr<Anchor>> anchors;
```
Comment (lincolnfrog) [inline third_party/tango/libtango_chromium/AnchorManager.h]: m_anchors?

File: third_party/tango/libtango_chromium/AnchorManager.h [hunk 49]
```diff
@@ -0,0 +1,66 @@
...
+                                    const float* cameraModelMatrix,
+                                    const float* anchorModelMatrix);
+  void removeAnchor(uint32_t identifier);
+  void removeAllAnchors();
+
+  /**
+  * Return: The list of updated anchors.
+  */
+  std::vector<std::shared_ptr<Anchor>> update(
+    double newTimestamp, const float* newCameraModelMatrix);
```
Comment (lincolnfrog) [inline third_party/tango/libtango_chromium/AnchorManager.h]: I think this should be 4 spaces (continuation indent)

File: third_party/tango/libtango_chromium/TangoHandler.h [hunk 50]
```diff
@@ -64,6 +61,14 @@ class PlaneDeltas
 	std::vector<long> removed;
 };
 
+class TangoHandlerEventListener
+{
```
Comment (lincolnfrog) [inline third_party/tango/libtango_chromium/TangoHandler.h]: { should be on the previous line according to Google c++ style (go/cstyle)

File: third_party/tango/libtango_chromium/TangoHandler.h [hunk 51]
```diff
@@ -64,6 +61,14 @@ class PlaneDeltas
...
+public:
+  virtual void anchorsUpdated(
+  	const std::vector<std::shared_ptr<Anchor>>& anchors) = 0;
```
Comment (lincolnfrog) [inline third_party/tango/libtango_chromium/TangoHandler.h]: this is a tab

File: third_party/tango/libtango_chromium/TangoHandler.h [hunk 52]
```diff
@@ -77,7 +82,8 @@ class TangoHandler {
 
   ~TangoHandler();
 
-  void onCreate(JNIEnv* env, jobject activity, int activityOrientation, int sensorOrientation);
+  void onCreate(JNIEnv* env, jobject activity, 
+  							int activityOrientation, int sensorOrientation);
```
Comment (lincolnfrog) [inline third_party/tango/libtango_chromium/TangoHandler.h]: these are tabs

File: third_party/tango/libtango_chromium/TangoHandler.h [hunk 53]
```diff
@@ -89,9 +95,14 @@ class TangoHandler {
   bool getProjectionMatrix(float near, float far, float* projectionMatrix);
   bool hitTest(float x, float y, std::vector<Hit>& hits);
   bool getPlaneDeltas(PlaneDeltas& planeDeltas);
+  std::shared_ptr<Anchor> createAnchor(
+  	const float* anchorModelMatrix);
```
Comment (lincolnfrog) [inline third_party/tango/libtango_chromium/TangoHandler.h]: this is a tab - I don't think this even needs to be two lines

**Review by lincolnfrog — CHANGES_REQUESTED** [review]
Lots of style issues. I think we should fix but up to you how strict we want to be

**Comment by judax** [comment]
Added a patch to fix these styling issues.

**Review by lincolnfrog — APPROVED** [review]

**Pull request merged**